publicip/fetcher.go

```diff
@@ -0,0 +1,83 @@
+package publicip
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-traceroute/log"
+	"github.com/cenkalti/backoff/v5"
+)
+
+const ipCheckerCallTimeout = 2 * time.Second
+
+// ipCheckers list of reliable public IP checkers
+var ipCheckers = []string{
+	"https://icanhazip.com/",         // owned by cloudflare
+	"https://ipinfo.io/ip",           // same as our GeoIP info provider
+	"https://checkip.amazonaws.com/", // Amazon
+	"https://api.ipify.org/",         // Dedicated Public IP info and GeoIP info provider
+	"https://whatismyip.akamai.com/", // Akamai is a CDN Provider
+}
+
+func GetPublicIP(ctx context.Context, client *http.Client, backoffPolicy *backoff.ExponentialBackOff) (net.IP, error) {
+	for _, ipChecker := range ipCheckers {
+		ip, err := getPublicIPUsingIPChecker(ctx, client, backoffPolicy, ipChecker)
+		if err != nil {
+			log.Debugf("error fetching: %s, %s\n", ipChecker, err.Error())
+			continue
+		}
+		return ip, nil
+	}
+	return nil, errors.New("no IP found")
+}
+
+func getPublicIPUsingIPChecker(ctx context.Context, client *http.Client, backoffPolicy *backoff.ExponentialBackOff, dest string) (net.IP, error) {
+	req, err := http.NewRequest("GET", dest, nil)
+	if err != nil {
+		return nil, errors.New("failed to create new request: " + err.Error())
+	}
+
+	operation := func() (net.IP, error) {
+		return handleRequest(client, req)
+	}
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, ipCheckerCallTimeout)
+	defer cancel()
+	result, err := backoff.Retry(ctxWithTimeout, operation, backoff.WithBackOff(backoffPolicy))
```

Comment:
With 5 separate sources for gathering IP data, I'm worried that gathering the IP could take a lot of time with an exponential backoff.

Let's say even 2 of the 5 sources are down due to a large CP outage, simply failing fast and moving on to the next source would seem like a better idea